File: Geeks4Geeks/CoinChange.cpp
#include<bits/stdc++.h>
using namespace std;

int coinChange(int S[], int m, int n)
{
    int mat[n+1][m];
    for(int i=0;i<=n;i++)
        for(int j=0;j<m;j++)
            if(i==0) mat[i][j]=1;
            else
            {
                int x=(i>=S[j])? mat[i-S[j]][j] : 0;
                int y=(j>0) ? mat[i][j-1] : 0;
                mat[i][j]=x+y;
            }
    return mat[n][m-1];
}

int main()
{
    int t,n,m;
    cin>>t;
    while(t--)
    {
        cin>>m;
        int S[m];
        for(int i=0;i<m;i++) cin>>S[i];
        cin>>n;
        cout<<coinChange(S,m,n)<<endl;
    }
}

File: HackerRank/cracking the interview/coinchange.py
def coinchange(coins,m,n):
	mat=[]				#mat[i][j] represents the number of ways to make a change of amount 'i' using 'j+1' elements of the 'coins' array.	
	
	for _ in range(n+1):	#initialization
		mat.append([None]*(m))
	
	for i in range(n+1):
		for j in range(m):
			if i==0:
				mat[i][j]=1	#first row is all 1s as there is 0 amount. just one way to do that.
			else:
				x=mat[i-coins[j]][j] if i>=coins[j] else 0	#no. of ways to make change for 'i-coins[j]' dollars with 'j+1' type of coins.
				y=mat[i][j-1] if j>0 else 0					#no. of ways to make change for 'i' dollars with 'j' type of coins.
				mat[i][j]=x+y

	return mat[n][m-1]		#last entry of the matrix represents the final answer, i.e, number of ways to make change for n dollars using m type of coins.


n,m=[int(x) for x in input().split()]
coins=[int(x) for x in input().split()]
coins.sort()
print(coinchange(coins,m,n))